Mathlib/Data/Matroid/Rank/ENat.lean

```diff
@@ -351,6 +352,188 @@ lemma IsRkFinite.eRk_lt_top (h : M.IsRkFinite X) : M.eRk X < ⊤ :=
 
 @[deprecated (since := "2025-04-13")] alias eRk_lt_top_of_finite := IsRkFinite.eRk_lt_top
 
+/-- If `X` is a finite-rank-set, and `I` is a subset of `X` of cardinality
+no larger than the rank of `X` that spans `X`, then `I` is a basis for `X`. -/
+lemma IsRkFinite.isBasis_of_subset_closure_of_subset_of_encard_le (hX : M.IsRkFinite X)
+    (hXI : X ⊆ M.closure I) (hIX : I ⊆ X) (hI : I.encard ≤ M.eRk X) : M.IsBasis I X := by
+  obtain ⟨J, hJ⟩ := M.exists_isBasis (I ∩ M.E)
+  have hIJ := hJ.subset.trans inter_subset_left
+  rw [← closure_inter_ground] at hXI
+  replace hXI := hXI.trans <| M.closure_subset_closure_of_subset_closure hJ.subset_closure
+  have hJX := hJ.indep.isBasis_of_subset_of_subset_closure (hIJ.trans hIX) hXI
+  rw [← hJX.encard_eq_eRk] at hI
+  rwa [← Finite.eq_of_subset_of_encard_le (hX.finite_of_isBasis hJX) hIJ hI]
+
+/-- If `X` is a finite-rank set, and `Y` is a superset of `X` of rank no larger than that of `X`,
+then `X` and `Y` have the same closure. -/
+lemma IsRkFinite.closure_eq_closure_of_subset_of_eRk_ge_eRk (hX : M.IsRkFinite X) (hXY : X ⊆ Y)
+    (hr : M.eRk Y ≤ M.eRk X) : M.closure X = M.closure Y := by
+  obtain ⟨I, hI⟩ := M.exists_isBasis' X
+  obtain ⟨J, hJ, hIJ⟩ := hI.indep.subset_isBasis'_of_subset (hI.subset.trans hXY)
+  rw [hI.eRk_eq_encard, hJ.eRk_eq_encard] at hr
+  rw [← closure_inter_ground, ← M.closure_inter_ground Y,
+    ← hI.isBasis_inter_ground.closure_eq_closure,
+    ← hJ.isBasis_inter_ground.closure_eq_closure, Finite.eq_of_subset_of_encard_le
+      (hI.indep.finite_of_subset_isRkFinite hI.subset hX) hIJ hr]
+
+/-! ### Insertion -/
+
+lemma eRk_insert_le_add_one (M : Matroid α) (e : α) (X : Set α) :
+    M.eRk (insert e X) ≤ M.eRk X + 1 :=
+  union_singleton ▸ (M.eRk_union_le_eRk_add_eRk _ _).trans <|
+    add_le_add_left (by simpa using M.eRk_le_encard {e}) _
+
+lemma eRk_insert_eq_add_one (he : e ∈ M.E \ M.closure X) : M.eRk (insert e X) = M.eRk X + 1 := by
+  obtain ⟨I, hI⟩ := M.exists_isBasis' X
+  rw [← hI.closure_eq_closure, mem_diff, hI.indep.mem_closure_iff', not_and] at he
+  rw [← eRk_closure_eq, ← closure_insert_congr_right hI.closure_eq_closure, hI.eRk_eq_encard,
+    eRk_closure_eq, Indep.eRk_eq_encard (by tauto), encard_insert_of_not_mem (by tauto)]
+
+lemma exists_eRk_insert_eq_add_one_of_lt (h : M.eRk X < M.eRk Z) :
+    ∃ z ∈ Z \ X, M.eRk (insert z X) = M.eRk X + 1 := by
```

Comment:
I think this is the only lemma that uses `Z`, but you already have `variable Y`, it seems cleaner to just use that?